public/sw.js

```diff
@@ -0,0 +1,231 @@
+    return;
+  }
+  pollNotifications();
+}
+
+async function pollNotifications() {
+  if (!sessionId || Notification.permission !== "granted") {
+    return;
+  }
+
+  const canSendNotification = await shouldSendNotification();
+  if (!canSendNotification) {
+    console.log("Notification was recently shown. Skipping notification.");
+    notificationTimer = setTimeout(pollNotifications, 5000);
+    return;
+  }
+  try {
+    // Temporary json file/endpoint until I've received new endpoint
+    // const response = await fetch(`https://api.zeeguu.org/available_languages`);
+    const response = await fetch(`/mock_notification.json`);
+    if (response.ok) {
+      const data = await response.json();
+      const randomNumber = Math.floor(Math.random() * 4);
+      const notificationUrl = data[randomNumber].url
+        ? data[randomNumber].url
+        : "/articles";
+      if (data) {
+        self.registration.showNotification("Zeeguu Notification", {
+          body: data[randomNumber].message,
+          icon: "/logo192.png",
+          tag: "zeeguu-notification",
+          data: { url: notificationUrl },
+        });
+        await updateLastNotificationTime();
+      }
+    }
+  } catch (error) {
+    console.error("Error polling notifications:", error);
+  }
+
+  //Receive notification every 5 seconds (5000 as setTimeout uses milliseconds)
+  notificationTimer = setTimeout(pollNotifications, 5000);
+}
+
+async function shouldSendNotification() {
+  try {
+    const cache = await caches.open(NOTIFICATION_CACHE_NAME);
+    const responseNotification = await cache.match("lastNotificationTime");
+    const responseInteraction = await cache.match("lastInteractionTime");
+    const lastNotificationTime = responseNotification
+      ? await responseNotification.json()
+      : null;
+    const lastInteractionTime = responseInteraction
+      ? await responseInteraction.json()
+      : null;
+
+    const notificationThreshold = 5000;
+    const interactionThreshold = 10000;
+
+    const canSendNotification =
+      (!lastInteractionTime ||
+        Date.now() - lastInteractionTime >= interactionThreshold) &&
+      (!lastNotificationTime ||
+        Date.now() - lastNotificationTime >= notificationThreshold);
+    return canSendNotification;
+  } catch (error) {
+    console.error("Error checking notification time from cache:", error);
+    return true;
+  }
+}
+
+async function updateLastNotificationTime() {
+  try {
+    const cache = await caches.open(NOTIFICATION_CACHE_NAME);
+    const response = new Response(JSON.stringify(Date.now()));
+    await cache.put("lastNotificationTime", response);
+  } catch (error) {
+    console.error("Error updating notification time in cache:", error);
+  }
+}
+
+self.addEventListener("notificationclick", (event) => {
+  event.notification.close();
+  console.log("Notification clicked");
+  if (notificationTimer) {
+    clearTimeout(notificationTimer);
+    notificationTimer = null;
+  }
+  event.waitUntil(
```

Comment:
"The ExtendableEvent.waitUntil() method tells the event dispatcher that work is ongoing. It can also be used to detect whether that work was successful. In service workers, waitUntil() tells the browser that work is ongoing until the promise settles, and it shouldn't terminate the service worker if it wants that work to complete."

https://developer.mozilla.org/en-US/docs/Web/API/ExtendableEvent/waitUntil

In this scenario, the event.waitUntil makes sure that the service worker waits with terminating whenever a notification is clicked, so the service worker can complete some of the asynchronous actions such as finding and matching clients (windows) and navigating to the URL before it terminates.

If we didn't do this, the service worker might terminate too early and these tasks might not be completed and the notification click won't do as intended.